blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,734 @@
+has elements indexable by 1 and 2. (Those `arrayref` statements enforce
+bounds-checking, and ensure that Julia will throw an appropriate error if you
+call `applyf([100])`.)
+
+If you pass a `Vector{Bool}`, it will compile `applyf` again, this time specializing it for `Bool` elements:
+
+```julia-repl
+julia> @code_typed applyf([true,false])
+CodeInfo(
+1 ─     Base.arrayref(true, container, 1)::Bool
+│       Base.arrayref(true, container, 2)::Bool
+└──     return 4
+) => Int64
+```
+
+In this case, you can see that Julia knew those two `arrayref` statements would
+return a `Bool`, and since it knows the value of `f(::Bool)` it just went
+ahead and computed the result at compile time for you.
+
+After calling `applyf` with both sets of arguments, hidden away in Julia's "method cache" there will
+be two `MethodInstance`s of `applyf`, one specialized for `Vector{Int}` and the other specialized for `Vector{Bool}`.
+You don't normally see these, but Julia manages them for you; anytime you write
+code that calls `applyf`, it checks to see if this previous compilation work can be reused.
+
+For the purpose of this blog post, things start to get especially interesting if use a container that can store elements with different types (here, type `Any`):
+
+```julia-repl
+julia> c = Any[1, false];
+
+julia> applyf(c)
+3
+
+julia> @code_typed applyf(c)
+CodeInfo(
+1 ── %1  = Base.arrayref(true, container, 1)::Any
+│    %2  = (isa)(%1, Bool)::Bool
+└───       goto #3 if not %2
+2 ──       goto #6
+3 ── %5  = (isa)(%1, Int64)::Bool
+└───       goto #5 if not %5
+4 ──       goto #6
+5 ── %8  = Main.f(%1)::Int64
+└───       goto #6
+6 ┄─ %10 = φ (#2 => 2, #4 => 1, #5 => %8)::Int64
+│    %11 = Base.arrayref(true, container, 2)::Any
+│    %12 = (isa)(%11, Bool)::Bool
+└───       goto #8 if not %12
+7 ──       goto #11
+8 ── %15 = (isa)(%11, Int64)::Bool
+└───       goto #10 if not %15
+9 ──       goto #11
+10 ─ %18 = Main.f(%11)::Int64
+└───       goto #11
+11 ┄ %20 = φ (#7 => 2, #9 => 1, #10 => %18)::Int64
+│    %21 = Base.add_int(%10, %20)::Int64
+└───       return %21
+) => Int64
+```
+
+This may seem a lot more complicated, but the take-home message is actually
+quite simple. First, look at those `arrayref` statements: they are annotated
+`Any`, meaning that Julia can't predict in advance what they will return.
+Immediately after each reference, notice that there are two `isa` statements
+followed by a `ϕ`; all of this is essentially equivalent to
+
+```
+if isa(x, Bool)
+    value = 2
+elseif isa(x, Int)
+    value = 1
+else
+    value = f(x)::Int
+end
+```
+
+This is [union-splitting], a performance optimization for cases where an object might be of one of several types.
+
+### Triggering method invalidation
+
+One point in particular is worth noting: what's up with that final `f(x)::Int` call?
+Didn't it already handle all the possibilities?
+Well, all the possibilities *so far*.
```

Comment:
There's something odd about this: the world age in which `applyf` was compiled didn't have any other possibilities for `f(x)`. So isn't hitting this fallback always going to result in a world age error?

Reply:
It does give you a good MethodError if you pass in a type that it doesn't work with.